File: binary_tree/isValidBST.cpp
// leetcode 98. isValidBST cpp

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    bool isValidBST(TreeNode* root) 
    {
        return this->dfs(root, nullptr, nullptr);
    }

private:    
    bool dfs(TreeNode* root, TreeNode* maximum, TreeNode* minimum)
    {
        if(!root)
        {
            return true;
        }
        if(maximum != nullptr && root->val >= maximum->val)
        {
            return false;
        }
        if(minimum != nullptr && root->val <= minimum->val)
        {
            return false;
        }
        return this->dfs(root->left, root, minimum) && this->dfs(root->right, maximum, root);
    }
};

File: leetcode_hot_100/isValidBST.py
# hot100-lc 98. 验证二叉搜索树

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution(object):
    def isValidBST(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        return self.dfs(root, float('-inf'), float('inf'))

    def dfs(self, root, minimum, maximum):
        if not root:
            return True
        return minimum < root.val < maximum and self.dfs(root.left, minimum, root.val) and self.dfs(root.right, root.val, maximum)
